app/contacts/contact_dns.py

```diff
@@ -428,10 +442,59 @@ async def _generate_response_for_completed_message(self, message_id, dns_request
+            elif msg.message_type == self.MessageType.FileUploadRequest:
+                return self._process_upload_request(request_context)
+            elif msg.message_type == self.MessageType.FileUploadData:
+                return await self._process_upload_data(request_context)
             else:
                 self.log.warning('Unsupported message type %s' % msg.message_type.value)
                 return self._generate_nxdomain_response(dns_request_packet)
 
+    def _process_upload_request(self, request_context):
+        upload_metadata = self._unpack_json(request_context.request_contents)
+        if upload_metadata:
+            filename = upload_metadata.get('file')
+            requesting_paw = upload_metadata.get('paw')
+            if filename and requesting_paw:
+                self.log.debug('Received upload request for file %s for request ID %s' %
+                               (filename, request_context.request_id))
+                self.pending_uploads[request_context.request_id] = self.FileUploadRequest(
+                    request_context.request_id,
+                    requesting_paw,
+                    filename
+                )
+                return self._generate_server_ready_ipv4_response(request_context.dns_request)
+            else:
+                self.log.warning('Client file upload request (ID %s) is missing filename and/or paw' %
+                                 request_context.request_id)
+        else:
+            self.log.warning('Empty upload request received from message ID %s' % request_context.request_id)
+        return self._generate_nxdomain_response(request_context.dns_request)
+
+    async def _process_upload_data(self, request_context):
+        # Make sure we are expecting this upload
+        upload_request = self.pending_uploads.get(request_context.request_id)
+        if upload_request:
+            # Append the request ID to the filename to help make it unique
+            unique_filename = '-'.join([upload_request.filename, request_context.request_id])
+
+            # request_context.request_contents contains the file upload data
+            await self._submit_uploaded_file(upload_request.requesting_paw,
+                                             unique_filename,
+                                             request_context.request_contents)
+            return self._generate_server_ready_ipv4_response(request_context.dns_request)
+        else:
+            self.log.warning('Client sent upload data without first making an upload request (request ID %s)' %
+                             request_context.request_id)
+        return self._generate_nxdomain_response(request_context.dns_request)
+
+    async def _submit_uploaded_file(self, paw, filename, data):
+        if paw and filename and data:
+            created_dir = os.path.normpath('/' + paw).lstrip('/')
```

Comment:
should we follow the same naming pattern as in sandcat for upload path `paw-host`?
obviously we would have to pull the agent object and get the hostname from there, so it would be extra overhead.

Reply:
I'm happy to follow the same naming pattern. It's no big lift to have the agent send the hostname info in the initial upload request anyway. Makes sense to keep it consistent, so i'll add it in

Reply:
fixed